Dockerfile

```diff
@@ -0,0 +1,88 @@
+#
+# Stage: Composer install for production
+#
+FROM composer:1.7.3 AS composer
+
+COPY composer.json \
+    composer.lock \
+    symfony.lock \
+    ./
+
+RUN composer --no-interaction install --no-dev --ignore-platform-reqs --no-autoloader --no-suggest --prefer-dist
+
+COPY src/ src/
+
+RUN composer --no-interaction dump-autoload --classmap-authoritative
+
+
+
+#
+# Stage: Composer install for development
+#
+FROM composer AS composer-dev
+
+RUN composer --no-interaction install --ignore-platform-reqs --no-suggest --prefer-dist
+
+
+
+#
+# Stage: Production environment
+#
+FROM php:7.2.11-fpm-alpine as prod
+
+WORKDIR /app
+
+RUN mkdir var && \
+    chown www-data:www-data var
+
+RUN docker-php-ext-install \
+    opcache
+
+COPY LICENSE LICENSE
+COPY .docker/php.ini ${PHP_INI_DIR}/conf.d/00-app.ini
+COPY bin/ bin/
+COPY src/ src/
+COPY public/ public/
+COPY config/ config/
+COPY --from=composer /app/vendor/ vendor/
+
+USER www-data
+
+
+
+#
+# Stage: Test environment
+#
+FROM prod as test
+
+USER root
+
+RUN touch .phpcs-cache && \
+    chown www-data:www-data .phpcs-cache
+COPY tests/ tests/
+COPY phpcs.xml.dist \
+    phpstan.neon.dist \
+    phpunit.xml.dist \
+    ./
+COPY --from=composer-dev /app/vendor/ vendor/
+
+USER www-data
+
+
+
+#
+# Stage: Development environment
+#
+FROM test as dev
+
+USER root
+
+COPY --from=composer /usr/bin/composer /usr/bin/composer
+COPY composer.json \
+    composer.lock \
+    symfony.lock \
+    ./
+
+RUN chown --recursive www-data:www-data .
```

Comment:
`www-data` ownership of files in in general unsafe but this is only the `dev` version. CI would use the `test` target so it should find out about any idiosyncrasies of permissions/ownership during test runs

Reply:
Originally went for everything as `root` (since most of it already is), but as PHP-FPM runs as `www-data` is then can't write things (unless `var` were to be kept as `www-data:www-data` probably).

Reply:
What should it write? I am seeing this only done onto the `dev` target so it seems nothing that would be used in a `prod` environment?

Reply:
Hmm, currently just the cache, but after a bit of playing not been able to recreate the same problem. Executing `docker-compose exec app bin/console cache:clear` as `root` means `var` is a mixture of `www-data` and `root, but both users can write to it. Will revert and see if it can ever be recreated.